chain/store/store.go

```diff
@@ -367,49 +367,32 @@ func (cs *ChainStore) UnmarkBlockAsValidated(ctx context.Context, blkid cid.Cid)
 func (cs *ChainStore) SetGenesis(ctx context.Context, b *types.BlockHeader) error {
 	ts, err := types.NewTipSet([]*types.BlockHeader{b})
 	if err != nil {
-		return err
+		return xerrors.Errorf("failed to construct genesis tipset: %w", err)
 	}
 
-	if err := cs.PutTipSet(ctx, ts); err != nil {
-		return err
-	}
-
-	return cs.metadataDs.Put(ctx, dstore.NewKey("0"), b.Cid().Bytes())
-}
-
-func (cs *ChainStore) PutTipSet(ctx context.Context, ts *types.TipSet) error {
 	if err := cs.PersistTipsets(ctx, []*types.TipSet{ts}); err != nil {
-		return xerrors.Errorf("failed to persist tipset: %w", err)
+		return xerrors.Errorf("failed to persist genesis tipset: %w", err)
 	}
 
-	expanded, err := cs.expandTipset(ctx, ts.Blocks()[0])
-	if err != nil {
-		return xerrors.Errorf("errored while expanding tipset: %w", err)
+	if err := cs.AddToTipSetTracker(ctx, b); err != nil {
+		return xerrors.Errorf("failed to add genesis tipset to tracker: %w", err)
 	}
 
-	if expanded.Key() != ts.Key() {
-		log.Debugf("expanded %s into %s\n", ts.Cids(), expanded.Cids())
-
-		tsBlk, err := expanded.Key().ToStorageBlock()
-		if err != nil {
-			return xerrors.Errorf("failed to get tipset key block: %w", err)
-		}
-
-		if err = cs.chainLocalBlockstore.Put(ctx, tsBlk); err != nil {
-			return xerrors.Errorf("failed to put tipset key block: %w", err)
-		}
+	if err := cs.RefreshHeaviestTipSet(ctx, ts.Height()); err != nil {
```

Comment:
Hm... for genesis, are we sure any of this is right? Shouldn't we _force_ replace everything? Like:

1. Discard all tipsets.
2. Replace the first tipset.
3. Move on?

Otherwise, e.g., the `metadataDs.Put` line below could do the wrong thing.

(not like this is likely to cause issues, but it's still a bit funky)

Reply:
I think you're right, but I was going for matching existing behaviour here, which this should do?

Reply:
Yeah, you're right.

Reply:
Let's just keep a list of TODOs, merge this, then we can make followup PRs (if/when we get time).